File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int n=heights.size();
        stack<pair<int,int>> st;
        int maxSize=0;
        for(int i=0;i<n;i++){
            int start=i;
            while(!st.empty() && st.top().first>heights[i]){
                auto [h,ind]=st.top();
                maxSize=max(maxSize,h*(i-ind));
                st.pop();
                start=ind;
            }
            st.push({heights[i],start});
        }
        while(!st.empty()){
            auto [h,ind]=st.top();
            st.pop();
            maxSize=max(maxSize,h*(n-ind));
        }
        return maxSize;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.java
class Solution {
    public int largestRectangleArea(int[] heights) {
        int N=heights.length;
        Stack<Integer> st=new Stack<>();
        int max=0;
        for(int i=0;i<N;i++){
            while(!st.isEmpty() && heights[st.peek()] >= heights[i]){
                int curr=heights[st.pop()]*(st.isEmpty()? i: i-st.peek()-1);
                max=Math.max(max,curr);
            }
            st.push(i);
        }
        while(!st.isEmpty()){
            int curr=heights[st.pop()]*(st.isEmpty()? N:N-1-st.peek());
            max=Math.max(max,curr);
        }
        return max;
    }
}